src/stable_marriage.rs

```diff
@@ -0,0 +1,164 @@
+struct Individual {
+    cnt_match: Option<usize>,
+
+    // For efficiency: most preferred last for men, most preferred first for women
+    is_man: bool,
+    prefs: Vec<usize>,
+}
+
+impl Individual {
+    fn get_pref_idx(&self, idx: usize) -> Option<usize> {
+        if self.is_man {
+            self.prefs.iter().rev().position(|i| *i == idx)
+        } else {
+            self.prefs.iter().position(|i| *i == idx)
+        }
+    }
+
+    fn pop_preferred_if_unmatched(&mut self) -> Option<usize> {
+        if self.cnt_match.is_some() {
+            None
+        } else {
+            self.prefs.pop()
+        }
+    }
+}
+
+fn make_empty_matching(v: Vec<Vec<usize>>, is_man: bool) -> Vec<Individual> {
+    v.into_iter()
+        .map(|prefs|
+            Individual {
+                cnt_match: None,
+                is_man: is_man,
+                prefs: prefs,
+            }
+        )
+        .collect()
+}
+
+fn init_matching(
+    men: Vec<Vec<usize>>,
+    women: Vec<Vec<usize>>,
+) -> (Vec<Individual>, Vec<Individual>) {
+    let mut men = make_empty_matching(men, true);
+    let women = make_empty_matching(women, false);
+    for mut x in &mut men {
+        x.prefs.reverse();
+    }
+
+    (men, women)
+}
+
+fn extract_matching(v: &Vec<Individual>) -> Vec<Option<usize>> {
```

Comment:
Given how it's used, I think it could consume the `Vec`, like `v: Vec<Individual>`. This would “fix” the `.into_iter()` below, which is currently (I think) called on the implicitly-converted `&[Individual]` slice (instead of on the `Vec<Individual>`)